regex-parsing/src/main/java/org/sonarsource/analyzer/commons/regex/helpers/SimplifiedRegexCharacterClass.java

```diff
@@ -0,0 +1,367 @@
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import java.util.regex.Pattern;
+import javax.annotation.Nullable;
+import org.sonarsource.analyzer.commons.regex.ast.AutomatonState;
+import org.sonarsource.analyzer.commons.regex.ast.CharacterClassElementTree;
+import org.sonarsource.analyzer.commons.regex.ast.CharacterClassIntersectionTree;
+import org.sonarsource.analyzer.commons.regex.ast.CharacterClassTree;
+import org.sonarsource.analyzer.commons.regex.ast.CharacterRangeTree;
+import org.sonarsource.analyzer.commons.regex.ast.CharacterTree;
+import org.sonarsource.analyzer.commons.regex.ast.DotTree;
+import org.sonarsource.analyzer.commons.regex.ast.EscapedCharacterClassTree;
+import org.sonarsource.analyzer.commons.regex.ast.MiscEscapeSequenceTree;
+import org.sonarsource.analyzer.commons.regex.ast.RegexBaseVisitor;
+import org.sonarsource.analyzer.commons.regex.ast.RegexSyntaxElement;
+
+public class SimplifiedRegexCharacterClass {
+
+  /**
+   * This map defines the contents of the character class in the following way:<br>
+   * For any entry {@code codepoint -> tree}, all the codepoints from {@code codepoint} up to (and excluding) the next
+   * entry are in the character class and belong to the given tree.<br>
+   * For any entry {@code codepoint -> null}, all the codepoints from {@code codepoint} up to (and excluding) the next
+   * entry are not part of the character class.<br>
+   * So a codepoint is contained in this class if and only if {@code contents.floorEntry(codePoint).getValue()} is
+   * non-null and the tree returned by {@code getValue} will be the element of the character class which matches that
+   * code point.
+   */
+  private TreeMap<Integer, RegexSyntaxElement> contents = new TreeMap<>();
+
+  private boolean containsUnknownCharacters = false;
+
+  public SimplifiedRegexCharacterClass() {
+  }
+
+  public SimplifiedRegexCharacterClass(CharacterClassElementTree tree) {
+    add(tree);
+  }
+
+  public SimplifiedRegexCharacterClass(DotTree tree) {
+    add(tree);
+  }
+
+  @Nullable
+  public static SimplifiedRegexCharacterClass of(AutomatonState tree) {
+    if (tree instanceof CharacterClassElementTree) {
+      return new SimplifiedRegexCharacterClass((CharacterClassElementTree) tree);
+    } else if (tree instanceof DotTree) {
+      return new SimplifiedRegexCharacterClass((DotTree) tree);
+    } else {
+      return null;
+    }
+  }
+
+  public boolean isEmpty() {
+    return contents.isEmpty() && !containsUnknownCharacters;
+  }
+
+  public void add(CharacterClassElementTree tree) {
+    new Builder(this).visitInCharClass(tree);
+  }
+
+  public void add(DotTree tree) {
+    char[] orderedExcludedCharacters;
+    if (tree.activeFlags().contains(Pattern.DOTALL)) {
+      orderedExcludedCharacters = new char[] {};
+    } else if (tree.activeFlags().contains(Pattern.UNIX_LINES)) {
```

Comment:
This is another point where behavior may differ between languages. According to some quick testing, Python at least will consider `\n` to be the only line break character regardless of which flags are set (and doesn't seem to have an equivalent for the `UNIX_LINES` flag).